### ISSUE TITLE
Quadratics as NLconstraints

### ISSUE THREAD

**Issue by coreysharris** [JuliaOpt/JuMP.jl#1870]
(JuMP v18.5)

I am trying to solve a MIQCP with Bonmin.  The problem has several quadratic constraints, but JuMP tells me that I can not use `@constraint` because 
```
ERROR: LoadError: Solver does not support quadratic constraints
```

If I enter the quadratic expression "by hand", I can get it to work.  Like so:
```julia
using JuMP, AmplNLWriter
m = Model(solver=AmplNLSolver("bonmin"))

@variable(m, a, Int)
@variable(m, b, Int)
@variable(m, s >= 0 )
@variable(m, t >= 0 )

@NLconstraint(m, 2*x+(a+b)*y >= 0)
```
But I actually I have an array of linear forms in a and b, and I'd like to create many constraints from them.

My first try
```julia
M = [2 a-b; a+b 3]
v = [s t]

for j = 1:2
    @NLconstraint(m, sum(M[i,j]*v[i] for i=1:2) == 0)
end
```
fails at the @NLconstraint: Unexpected object 2 in nonlinear expression.  Ok, so I use the @NLexpression macro.  This fails on the first line (same error). 
```julia
M = @NLexpression(m, [i=1:2,j=1:2], [2 a-b; a+b 3][i,j])
v = @NLexpression(m, [i=1:2], [s t][i])

for j = 1:2
    @NLconstraint(m, sum(M[i,j]*v[i] for i=1:2) == 0)
end
```

What am I doing wrong?

**Comment by odow** [comment]
Reproducible on JuMP 0.19
```julia
model = Model()
@variable(model, x)
A = [x]
B = [2, x]
@NLconstraint(model, A[1] * B[1] == 0)
```

The problem is that `B = [2, x]` promotes `2` to an `AffExpr`, which are not allowed in `@NL` macros. Take a read of the [syntax notes](http://www.juliaopt.org/JuMP.jl/latest/nlp/#Syntax-notes-1).

A quick fix is to type `B` as `Any`.
```julia
model = Model()
@variable(model, x)
A = [x]
B = Any[2, x]
@NLconstraint(model, A[1] * B[1] == 0)
```

This needs an improved error message.

**Comment by coreysharris** [comment]
Ah, I see.  In my case, 
```julia
M = Any[2 a-b; a+b 3]
```
still generated the error at `a-b`, but that's obviously an AffExpr.  My solution (in case it's useful for others):
```julia
@variable(m, aux[1:2,1:2])
M = [2 a-b; a+b 3]
@constraint(m, aux .== M)
```

I created #1872 to address the error message.

**Comment by mlubin** [comment]
Closed by #1872. The underlying "quadratics not supported" issue is https://github.com/JuliaOpt/AmplNLWriter.jl/issues/49.
